core/src/broadcast_stage.rs

```diff
@@ -406,18 +431,21 @@ pub fn broadcast_shreds(
     peers: &[ContactInfo],
     last_datapoint_submit: &Arc<AtomicU64>,
     transmit_stats: &mut TransmitShredsStats,
+    leader_schedule_cache: &Arc<LeaderScheduleCache>,
+    bank_forks: &Arc<RwLock<BankForks>>,
 ) -> Result<()> {
     let broadcast_len = peers_and_stakes.len();
     if broadcast_len == 0 {
         update_peer_stats(1, 1, last_datapoint_submit);
         return Ok(());
     }
     let mut shred_select = Measure::start("shred_select");
+    let root_bank = bank_forks.read().unwrap().root_bank();
     let packets: Vec<_> = shreds
         .iter()
         .map(|shred| {
-            let broadcast_index = weighted_best(peers_and_stakes, shred.seed());
-
+            let seed = shred.seed(leader_schedule_cache, &root_bank);
+            let broadcast_index = weighted_best(peers_and_stakes, seed);
```

Comment:
I think here, the leader is the node itself which is broadcasting shreds. cc @carllin to confirm if this is always the case.

So we do not need `leader_schedule_cache`, and we can avoid `leader_pubkey` lookup in `Shred::seed`.
Instead this function should take a `self_pubkey: &Pubkey` argument, and pass it directly to `Shred::seed` as the `leader_pubkey`.

Reply:
updated